lib/propolis/src/lifecycle.rs

```diff
@@ -87,3 +89,96 @@ pub trait Lifecycle: Send + Sync + 'static {
         Migrator::Empty
     }
 }
+
+/// Indicator for tracking [Lifecycle] states.
+///
+/// As implementors of the [Lifecycle] trait are driven through various state
+/// changes when called through its functions, the underlying resource may want
+/// to keep track of which state it is in.  Device emulation may wish to
+/// postpone certain setup tasks when undergoing state import during a
+/// migration, for example.
+///
+/// The [Indicator] is meant as a convenience tool for such implementors to
+/// simply track the current state by calling the appropriate methods
+/// corresponding to their [Lifecycle] counterparts.
+#[derive(Default)]
+pub struct Indicator(AtomicUsize);
+impl Indicator {
+    pub const fn new() -> Self {
+        Self(AtomicUsize::new(IndicatedState::Init as usize))
+    }
+    /// Indicate that holder is has started
+    ///
+    /// To be called as part of the implementor's [Lifecycle::start()] method
+    pub fn start(&self) {
+        self.state_change(IndicatedState::Run);
+    }
+    /// Indicate that holder is has paused
+    ///
+    /// To be called as when [Lifecycle::pause()] has been called on the
+    /// implementor _and_ that pause has completed (if there is async work
+    /// required for such a state).
+    pub fn pause(&self) {
+        self.state_change(IndicatedState::Pause);
+    }
+    /// Indicate that holder is has resumed
+    ///
+    /// To be called as part of the implementor's [Lifecycle::resume()] method
+    pub fn resume(&self) {
+        self.state_change(IndicatedState::Run);
+    }
+    /// Indicate that holder is has halted
+    ///
+    /// To be called as part of the implementor's [Lifecycle::halt()] method
+    pub fn halt(&self) {
+        self.state_change(IndicatedState::Halt);
+    }
+    /// Get the currently indicated state for the holder
+    pub fn state(&self) -> IndicatedState {
+        IndicatedState::from_usize(self.0.load(Ordering::SeqCst))
+    }
+
+    fn state_change(&self, new: IndicatedState) {
+        let old = self.0.swap(new as usize, Ordering::SeqCst);
```

Comment:
I don't think sequentially-consistent ordering is required here. I don't believe there is another memory location that we want to have a total ordering of sequentially-consistent stores/loads with, so `AcqRel` should be sufficient here?

I doubt the performance penalty for sequentially-consistent stores matters a whole lot here, though, so :woman_shrugging: